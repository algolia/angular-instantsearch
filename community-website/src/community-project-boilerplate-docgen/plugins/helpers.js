const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
