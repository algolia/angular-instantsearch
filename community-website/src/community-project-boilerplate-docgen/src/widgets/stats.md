---
title: Widgets - Stats
layout: widget.pug
---

# Stats

## Description

The Stats component displays the total number of matching hits and the time it took to get them (time spent in the Algolia server).

## Code example

You can use the directive `<ng-template></ng-template>` to customize the output:

```js
@Component({
  selector: 'my-app',
  template: `
    <ais-stats>
      <ng-template let-state="state">
        {{state.nbHits}} results found in {{stats.processingTimeMS}}ms.
      </ng-template>
    </ais-stats>
  `
})
export class AppComponent {}
```

The `state` object contains:

* `hitsPerPage` The maximum number of hits per page returned by Algolia
* `nbHits` The number of hits in the result set
* `nbPages` The number of pages computed for the result set.
* `page` The current page.
* `processingTimeMS` The time taken to compute the results inside the Algolia engine.
* `query` The query used for the current search.
