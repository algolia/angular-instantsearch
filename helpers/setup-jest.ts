import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
