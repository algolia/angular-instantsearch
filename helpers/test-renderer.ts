import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
