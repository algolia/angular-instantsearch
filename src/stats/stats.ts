import {
  Component,
  ContentChild,
  TemplateRef,
  Inject,
  PLATFORM_ID
} from "@angular/core";
import { connectStats } from "instantsearch.js/es/connectors";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";

@Component({
  selector: "ng-ais-stats",
  template: `
    <div [class]="cx()">
      <ng-container *ngTemplateOutlet="template; context: templateContext">
      </ng-container>

      <span *ngIf="!template" [class]="cx('text')">
        {{state.nbHits}} results found in {{state.processingTimeMS}}ms.
      </span>
    </div>
  `
})
export class NgAisStats extends BaseWidget {
  @ContentChild(TemplateRef) public template: any;

  public state = {
    hitPerPage: 0,
    nbHits: 0,
    nbPages: 0,
    page: 0,
    processingTimeMS: 0,
    query: ""
  };

  get templateContext() {
    return { state: this.state };
  }

  constructor(
    @Inject(PLATFORM_ID) public platformId: Object,
    searchInstance: NgAisInstance
  ) {
    super(searchInstance, "Stats");
    this.createWidget(connectStats);
  }
}
