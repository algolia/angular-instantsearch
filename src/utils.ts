export function bem(widgetName: string) {
  const cx = function(element?: string, subElement?: string) {
    if (element) {
      const scoppedWidgetName = `ais-${widgetName}-${element}`;

      // output `ais-Widget-Header|Body|Footer ais-Header|Body|Footer`
      if (element === "header" || element === "body" || element === "footer") {
        const nonScoppedWidgetName = `ais-${element}`;
        return `${scoppedWidgetName} ${nonScoppedWidgetName}`;
      }

      // output `ais-Widget-Xyz--abc`
      if (subElement) {
        return `${scoppedWidgetName}--${subElement}`;
      }

      // output `ais-Widget-Xyz`
      return scoppedWidgetName;
    }

    // output `ais-Widget`
    return `ais-${widgetName}`;
  };
  return cx;
}

export function parseNumberInput(input?: number | string) {
  return typeof input === "string" ? parseInt(input, 10) : input;
}
