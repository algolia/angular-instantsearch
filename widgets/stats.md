## `<ng-ais-stats />`

The Stats component displays the total number of matching hits and the time it took to get them (time spent in the Algolia server).

| Props     | Type     | Description
| -         | -        | -
| `header?` | `string` | Displays text before widget
| `footer?` | `string` | Displays text after widget

You can use the directive `<ng-template></ng-template>` to customize the output:

```ts
@Component({
  selector: 'my-app',
  template: `
    <ng-ais-stats>
      <ng-template let-state="state">
        {{stats.nbHits}} results found in {{stats.processingTimeMS}}ms.
      </ng-template>
    </ng-ais-stats>
  `
})
export class AppComponent {}
```

The `state` object contains:

* `hitsPerPage` The maximum number of hits per page returned by Algolia
* `nbHits` The number of hits in the result set
* `nbPages` The number of pages computed for the result set.
* `page` The current page.
* `processingTimeMS` The time taken to compute the results inside the Algolia engine.
* `query` The query used for the current search.