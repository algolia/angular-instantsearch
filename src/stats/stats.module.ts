import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisStats } from './stats';

@NgModule({
  declarations: [NgAisStats],
  entryComponents: [NgAisStats],
  exports: [NgAisStats],
  imports: [CommonModule],
})
export class NgAisStatsModule {}
