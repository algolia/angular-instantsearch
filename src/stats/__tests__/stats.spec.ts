import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisStats } from "../stats";

import { bem } from "../../utils";

const cx = bem("Stats");

const render = createRenderer({
  defaultState: {
    hitPerPage: 20,
    nbHits: 100,
    nbPages: 5,
    page: 0,
    processingTimeMS: 123,
    query: "foobar"
  },
  template: "<ais-stats></ais-stats>",
  TestedWidget: NgAisStats
});

describe("Stats", () => {
  it("should render without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("should render with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });
});
