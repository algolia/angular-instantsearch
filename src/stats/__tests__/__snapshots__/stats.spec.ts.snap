// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats should render with state 1`] = `
<ng-component
  testedWidget={[Function NgAisStats]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-stats>
      
      
      <div
        class="ais-Stats"
      >
        
        
        
        
  
        
        
        <span
          class="ais-Stats-text"
        >
          
          100 results found in 123ms.
        
        </span>
        
      
      </div>
      
    
    </ng-ais-stats>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;

exports[`Stats should render without state 1`] = `
<ng-component
  testedWidget={[Function NgAisStats]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-stats>
      
      
      <div
        class="ais-Stats"
      >
        
        
        
        
  
        
        
        <span
          class="ais-Stats-text"
        >
          
          0 results found in 0ms.
        
        </span>
        
      
      </div>
      
    
    </ng-ais-stats>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;
